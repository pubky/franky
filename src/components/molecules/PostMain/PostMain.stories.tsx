import type { Meta, StoryObj } from '@storybook/nextjs-vite';
import { PostMain } from './PostMain';

const meta: Meta<typeof PostMain> = {
  title: 'Molecules/Post/PostMain',
  component: PostMain,
  parameters: {
    layout: 'centered',
    backgrounds: {
      default: 'dark',
      values: [
        { name: 'dark', value: '#05050A' },
        { name: 'light', value: '#F2F2F7' },
      ],
    },
  },
  args: {
    avatarSrc: '/images/avatar.png',
    displayName: 'Satoshi Nakamoto',
    label: '1YXP...7R32',
    timeLabel: '15m',
    text: 'I have said it before: If you donâ€™t believe me or donâ€™t get it, I donâ€™t have time to try to convince you.',
    tags: [
      { label: 'bitcoin', count: 21, color: '#FF9900' },
      { label: 'based', count: 7, color: '#FF0000' },
      { label: 'ðŸ‘€', count: 5, color: '#00F0FF' },
    ],
    showAddButton: true,
    showEmojiPicker: true,
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: (args) => (
    <div className="w-[720px]">
      <PostMain {...args} />
    </div>
  ),
};

export const Mobile: Story = {
  render: (args) => (
    <div className="w-[375px]">
      <PostMain {...args} />
    </div>
  ),
  parameters: {
    viewport: { defaultViewport: 'mobile1' },
    docs: {
      description: {
        story: 'Mobile layout with stacked Tags and Actions and tighter paddings.',
      },
    },
  },
};
