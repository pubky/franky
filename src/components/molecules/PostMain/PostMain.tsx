import * as React from 'react';
import * as Molecules from '@/molecules';
import * as Atoms from '@/atoms';
import * as Libs from '@/libs';

import type { PostTagsListTag } from '@/molecules/PostTagsList/PostTagsList';

export interface PostMainProps {
  className?: string;
  // Header
  avatarSrc?: string;
  avatarAlt?: string;
  displayName: string;
  label?: string;
  timeLabel?: string;
  // Content
  text: string;
  // Tags
  tags?: PostTagsListTag[];
  showAddButton?: boolean;
  showEmojiPicker?: boolean;
  showTagClose?: boolean;
  onTagClick?: (tag: PostTagsListTag, index: number) => void;
  onTagClose?: (tag: PostTagsListTag, index: number) => void;
  onTagAdd?: (label: string) => void;
  onEmojiClick?: () => void;
  // Actions
  tagCount?: number;
  replyCount?: number;
  repostCount?: number;
  isBookmarked?: boolean;
  onTagActionClick?: () => void;
  onReplyClick?: () => void;
  onRepostClick?: () => void;
  onBookmarkClick?: () => void;
  onMoreClick?: () => void;
}

export function PostMain({
  className,
  avatarSrc,
  avatarAlt,
  displayName,
  label,
  timeLabel,
  text,
  tags = [],
  showAddButton = true,
  showEmojiPicker = true,
  showTagClose = false,
  onTagClick,
  onTagClose,
  onTagAdd,
  onEmojiClick,
  tagCount = 5,
  replyCount = 7,
  repostCount = 3,
  isBookmarked = false,
  onTagActionClick,
  onReplyClick,
  onRepostClick,
  onBookmarkClick,
  onMoreClick,
}: PostMainProps) {
  return (
    <Atoms.Card className={Libs.cn('w-full', className)}>
      <Atoms.CardContent className="px-4 md:px-6 md:py-0">
        <div className="flex flex-col gap-4 md:gap-4 w-full">
          <Molecules.PostHeader
            avatarSrc={avatarSrc}
            avatarAlt={avatarAlt}
            displayName={displayName}
            label={label}
            timeLabel={timeLabel}
          />

          <Molecules.PostContent text={text} />

          <div className="flex flex-col md:flex-row items-stretch md:items-center gap-4  flex-wrap md:flex-nowrap justify-start md:justify-start">
            <Molecules.PostTagsList
              tags={tags}
              showInput={false}
              showAddButton={showAddButton}
              addMode
              showEmojiPicker={showEmojiPicker}
              showTagClose={showTagClose}
              onTagClick={onTagClick}
              onTagClose={onTagClose}
              onTagAdd={onTagAdd}
              onEmojiClick={onEmojiClick}
            />

            <Molecules.PostActionsBar
              tagCount={tagCount}
              replyCount={replyCount}
              repostCount={repostCount}
              isBookmarked={isBookmarked}
              onTagClick={onTagActionClick}
              onReplyClick={onReplyClick}
              onRepostClick={onRepostClick}
              onBookmarkClick={onBookmarkClick}
              onMoreClick={onMoreClick}
              className="justify-start w-full flex-1 md:justify-end"
            />
          </div>
        </div>
      </Atoms.CardContent>
    </Atoms.Card>
  );
}
