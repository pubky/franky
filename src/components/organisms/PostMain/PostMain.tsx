'use client';

import { useLiveQuery } from 'dexie-react-hooks';
import * as Core from '@/core';
import * as Libs from '@/libs';
import * as Atoms from '@/atoms';
import * as Molecules from '@/molecules';
import * as Organisms from '@/organisms';

export interface PostMainProps {
  postId: string;
  onClick?: () => void;
  className?: string;
}

export function PostMain({ postId, onClick, className }: PostMainProps) {
  // Fetch post tags
  const postTags = useLiveQuery(async () => {
    return await Core.PostTagsModel.findById(postId);
  }, [postId]);

  const tags =
    postTags?.tags.map((tag, index) => ({
      id: `${postId}-tag-${index}`,
      label: tag.label,
    })) || [];

  return (
    <div onClick={onClick} className="cursor-pointer">
      <Atoms.Card className={Libs.cn('w-full', className)}>
        <Atoms.CardContent className="px-4 md:px-6 md:py-0">
          <div className="flex flex-col gap-4 md:gap-4 w-full">
            <Organisms.PostHeader postId={postId} />

            <Organisms.PostContent postId={postId} />

            <div className="flex flex-col md:flex-row items-stretch md:items-center gap-4 flex-wrap md:flex-nowrap justify-start md:justify-start">
              <Molecules.PostTagsList
                tags={tags}
                showInput={false}
                showAddButton={false}
                addMode
                showEmojiPicker={false}
                showTagClose={false}
              />

              <Organisms.PostActionsBar postId={postId} className="justify-start w-full flex-1 md:justify-end" />
            </div>
          </div>
        </Atoms.CardContent>
      </Atoms.Card>
    </div>
  );
}
