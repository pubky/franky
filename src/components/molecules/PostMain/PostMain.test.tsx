import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { PostMain } from './PostMain';

vi.mock('@/molecules', async (importOriginal) => {
  const actual = await importOriginal<typeof import('@/molecules')>();
  return {
    ...actual,
    PostHeader: ({ displayName, label, timeLabel }: { displayName: string; label: string; timeLabel: string }) => (
      <div data-testid="header">
        <div>{displayName}</div>
        <div>{label}</div>
        <div>{timeLabel}</div>
      </div>
    ),
    PostContent: ({ text }: { text: string }) => <div data-testid="content">{text}</div>,
    PostTagsList: ({ tags }: { tags?: { label: string }[] }) => (
      <div data-testid="tags">{tags?.map((t: { label: string }) => t.label).join(',')}</div>
    ),
    PostActionsBar: ({
      tagCount,
      replyCount,
      repostCount,
    }: {
      tagCount?: number;
      replyCount?: number;
      repostCount?: number;
    }) => <div data-testid="actions">{`${tagCount}-${replyCount}-${repostCount}`}</div>,
  };
});

describe('PostMain', () => {
  it('renders all sections', () => {
    render(
      <PostMain
        displayName="Satoshi Nakamoto"
        label="1YXP...7R32"
        timeLabel="15m"
        text="Hello"
        tags={[{ label: 'bitcoin' }]}
      />,
    );
    expect(screen.getByTestId('header')).toBeInTheDocument();
    expect(screen.getByTestId('content')).toBeInTheDocument();
    expect(screen.getByTestId('tags')).toHaveTextContent('bitcoin');
    expect(screen.getByTestId('actions')).toHaveTextContent('5-7-3');
  });

  it('matches snapshot', () => {
    const { container } = render(
      <PostMain
        displayName="Satoshi Nakamoto"
        label="1YXP...7R32"
        timeLabel="15m"
        text="Hello"
        tags={[{ label: 'bitcoin' }]}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
