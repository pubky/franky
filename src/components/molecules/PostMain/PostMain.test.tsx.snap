// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostMain > matches snapshot 1`] = `
<div
  class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl py-6 shadow-sm w-full"
  data-slot="card"
  data-testid="card"
>
  <div
    class="px-4 md:px-6 md:py-0"
    data-slot="card-content"
  >
    <div
      class="flex flex-col gap-4 md:gap-4 w-full"
    >
      <div
        data-testid="header"
      >
        <div>
          Satoshi Nakamoto
        </div>
        <div>
          1YXP...7R32
        </div>
        <div>
          15m
        </div>
      </div>
      <div
        data-testid="content"
      >
        Hello
      </div>
      <div
        class="flex flex-col md:flex-row items-stretch md:items-center gap-4  flex-wrap md:flex-nowrap justify-start md:justify-start"
      >
        <div
          data-testid="tags"
        >
          bitcoin
        </div>
        <div
          data-testid="actions"
        >
          5-7-3
        </div>
      </div>
    </div>
  </div>
</div>
`;
